File: portscanner.py
import socket
import time
import sys
import threading
from queue import Queue
#Author: aKa slowkep
socket.setdefaulttimeout(1)
print_lock = threading.Lock()

if len(sys.argv) == 2:
	target = socket.gethostbyname(sys.argv[1])
else:
	print("Argumento invalido.")
	print("Syntax: python3 scanner.py <ip>")
	
def portscan(port):
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        con = s.connect((target, port))
        with print_lock:
            print('Port', port, 'is open!')
        con.close()
    except:
        pass

def threader():
    while True:
        worker = q.get()
        portscan(worker)
        q.task_done()

q = Queue()
startTime = time.time()

for x in range(100):
    t = threading.Thread(target=threader)
    t.daemon = True
    t.start()

for worker in range(1, 65536):
    q.put(worker)

q.join()

print('Time taken:', time.time() - startTime)

File: portscanner.rb
#!/usr/bin/ruby
# Ejemplo: ruby scanner.rb <target> > scan.txt |  grep "open" scan.txt
require 'socket'
require 'thread'
require 'timeout'
#Author: aKa slowkep

host = ARGV[0] #Le proprocionamos un argumento.

def scan(host) #Definimos
  (1..1000).each do |port| #Podemos modificar el rango de puertos, y especificamos para que cada puerto haga..:
    Thread.new {
      begin
        Timeout.timeout(0.001) do                    # tiempo de timeout
          s = TCPSocket.new(host, port)            # Crear nuevo socket despues del timeout
          puts "[+] #{host} | Port #{port} open"
          s.close                                  #Cerramos variable
        end
      rescue Errno::ECONNREFUSED    #En caso de puerto cerrado
        # puts "[!] #{host} | Port #{port} closed"
        next
      rescue Timeout::Error # En caso de que el puerto no de respuesta /filtered
        puts "[!] #{host} | Port #{port} timeout/filtered"
        next
      end
    }.join
  end
end

scan host
